test/unit/math/rev/functor/test_fixture_ode.hpp

```diff
@@ -0,0 +1,470 @@
+   * - param(): return to ODE vector-like parameters
+   * - init(): return ODE vector-like init condition
+   * - times(): return ODE vector time step
+   * - apply_solver(init, param) solve ode given vector-like init & param.
+   *   It must return data results when both init & param are data.
+   *
+   * @param diff finite diff stepsize
+   * @param tol double value test tolerance
+   */
+  void test_fd_vv(double diff, double tol) {
+    std::stringstream msgs;
+    ode_problem_type& ode = static_cast<ode_problem_type&>(*this);
+
+    stan::math::nested_rev_autodiff nested;
+
+    int n = ode.init().size();
+    int m = ode.param().size();
+    std::vector<std::vector<Eigen::VectorXd>> fd_res_y(n);
+    for (size_t i = 0; i < n; ++i) {
+      fd_res_y[i] = fd_init(i, diff);
+    }
+
+    std::vector<std::vector<Eigen::VectorXd>> fd_res_p(m);
+    for (size_t i = 0; i < m; ++i) {
+      fd_res_p[i] = fd_param(i, diff);
+    }
+
+    std::vector<stan::math::var> vars(
+        stan::math::to_array_1d(stan::math::to_var(ode.init())));
+    auto theta = ode.param();
+    for (int i = 0; i < m; ++i) {
+      vars.push_back(theta[i]);
+    }
+    Eigen::Matrix<stan::math::var, -1, 1> yv(n);
+    for (int i = 0; i < n; ++i) {
+      yv(i) = vars[i];
+    }
+    std::vector<stan::math::var> theta_v(vars.begin() + n, vars.end());
+
+    std::vector<Eigen::Matrix<stan::math::var, -1, 1>> ode_res
+        = ode.apply_solver(yv, theta_v);
+
+    std::vector<double> grads_eff;
+    for (size_t i = 0; i < ode.times().size(); i++) {
+      for (size_t j = 0; j < n; j++) {
+        grads_eff.clear();
+        ode_res[i][j].grad(vars, grads_eff);
+
+        for (size_t k = 0; k < m; k++) {
+          EXPECT_NEAR(grads_eff[k + n], fd_res_p[k][i][j], tol)
+              << "Gradient of ode solver failed with initial positions"
+              << " unknown and parameters unknown for param at time index " << i
+              << ", equation index " << j << ", and parameter index: " << k;
+        }
+        for (size_t k = 0; k < n; k++) {
+          EXPECT_NEAR(grads_eff[k], fd_res_y[k][i][j], tol)
+              << "Gradient of ode solver failed with initial positions"
+              << " unknown and parameters known for initial position at time "
+                 "index "
+              << i << ", equation index " << j
+              << ", and parameter index: " << k;
+        }
+
+        nested.set_zero_all_adjoints();
+      }
+    }
+  }
+
+  /**
+   * Test gradient w.r.t times.
+   *
+   * Require methods in child fixture:
+   * - times(): return ODE var vector time steps
+   * - apply_solver(): solve ODE assuming var time step
+   * - eval_rhs(double t, VectorXd x): eval ODE right-hand-side at
+   *   time t and independent variable x.
+   */
+  void test_ts_ad() {
+    ode_problem_type& ode = static_cast<ode_problem_type&>(*this);
+    std::vector<double> g;
+    std::vector<Eigen::Matrix<stan::math::var, -1, 1>> res = ode.apply_solver();
```

Comment:
Can this be using a nested environment as well (like the others)? Thanks!

Reply:
Here the intention is to allow test `var` `ts` combined with `double` or `var` theta & y0. It's up to the test model which combination to test. See https://github.com/stan-dev/math/blob/f1971b9cbf674715b4c49b12168d5b4d7c7631f6/test/unit/math/rev/functor/fho_ode_typed_ts_test.cpp#L18 So if we want to use `nested` it has to be included in the upper level, either in the child fixture or the parent fixture. I'd like to avoid to burden tester with details of `set_zero_adjoint` and also avoid to have `test_fixture_ode.hpp` to carry a state (nest variable). Hence the call to global set_zero_adjoint.